### DISCUSSION TITLE
I made an event emitter library for Teal

### DISCUSSION THREAD

**Discussion by mtdowling** [teal-language/tl#872] [Show and tell]
I made a somewhat unique strongly typed event emitter library for Teal/Lua: https://github.com/mtdowling/emitter.tl. I also [published it to Luarocks](https://luarocks.org/modules/mtdowling/emitter.tl) in way that includes the Lua and Teal code, [based on advice from a previous discussion](https://github.com/teal-language/tl/discussions/836).

```teal
local Emitter = require("emitter")
local emitter = Emitter.new()

local record WarningEvent is Emitter.Event
    message: string
end

function WarningEvent.new(message: string): WarningEvent
    return { type = WarningEvent, message = message }
end

emitter:on(WarningEvent, function(event: WarningEvent)
    print(event.message)
end)

emitter:emit(WarningEvent.new("Warning message"))
```

Like others in the discussion board, I thought I'd give some thoughts on the DX of using Teal, Cyan, LuaRocks, etc.

**Good stuff**

* The new interface feature of Teal feels like it was the missing piece that now makes defining types with Teal far more expressive.
* Once I setup Neovim with [teal-language-server](https://github.com/teal-language/teal-language-server) and [vim-teal](https://github.com/teal-language/vim-teal), I had a really nice editing experience.
* I was able to get Teal and Lua files bundled into my rock.

**Wishlist**

(1) The build tooling around Teal feels a bit disjointed when compared to things like Cargo, NPM, or Gradle. Those tools all build, test, and publish libraries. But with a Teal library, I found myself using Cyan+tlconfig.lua (to get the LSP to detect Teal), luarocks to install test and build dependencies, luarocks to publish the package, and Make to bundle all the commands up. I guess my hope is that Cyan will somehow orchestrate all of this and provide the same kind of all in one experience lots of other build tools and package managers offer. For example, to run the tests or a build, I want a tool that makes sure my dependencies are up to date with my LuaRocks file and then runs the right script. LuaRocks can partially do this, but it doesn't allow for multi-step builds without building a completely custom build command.

(2) Better `any` type integration with generics. If something takes a `Foo<any>` and I give it a `Foo<T>`, it should work.

```teal
local record Foo<T>
   value: T
end

local bar: Foo<number> = { value = 42 }

local function baz(v: Foo<any>)
    print(v.value)
end

baz(bar) -- <- type parameter <<any type>>: got number, expected <any type>
```

(3) I didn't dig in yet to figure out how I could easily write my busted tests in Teal, but that would be a great How-To article somewhere.

(4) There are a few other rough edges I ran into around recursive and/or "self" generics that I'll open issues for.

All in all, the "next" branch release of Teal feels really great and I'm hoping to open source a few other libraries for it soon.

**Comment by euclidianAce** [comment]
Thanks for the feedback! Main maintainer of cyan here.

> I guess my hope is that Cyan will somehow orchestrate all of this and provide the same kind of all in one experience lots of other build tools and package managers offer

I would like to have more Cyan <-> Luarocks integration in the future for the reasons you've stated above. If there are certain features you had in mind I'd appreciate if you opened some issues in the cyan repo.

> I didn't dig in yet to figure out how I could easily write my busted tests in Teal, but that would be a great How-To article somewhere.

Not an explicit tutorial, but I recently ported Cyan's tests to Teal and that could be a good jumping off point. The gist of it is that we have a `src/spec` that gets built into `build/spec` and running `busted build` will run the compiled tests. (Additionally the [teal-types repo](https://github.com/teal-language/teal-types) has type definitions for busted. It is still a two step process but is at least a jumping off point.

**Reply by mtdowling** [reply]
Awesome! Glad to hear there's interest in this direction. I opened a few issues on the Cyan repo around some of the things that would help streamline the DX and remove the need for make.

* [LuaRocks dependency integration](https://github.com/teal-language/cyan/issues/50) - install build and test dependencies automatically
* [Named script tasks](https://github.com/teal-language/cyan/issues/51) - no need to remember arguments and script names for builds, like NPM
* [Ability to run tests](https://github.com/teal-language/cyan/issues/52) - this would make testing with Busted+Teal seamless

**Reply by Aire-One** [reply]
Alternatively, about tests, I have pushed this rock some times ago https://luarocks.org/modules/aire-one/busted-tl.

Basically, it adds a new `module.file` to busted, so it can load teal files when using the `loaders=teal` parameter.

**Reply by mtdowling** [reply]
@Aire-One that's awesome! I was able to use this and migrate my tests to Teal. [I couldn't figure out](https://github.com/mtdowling/emitter.tl/commit/a26c630b41d3ad259a1a1e3301589e4727a13eeb) how to specify the dependency for this in my rockspec since there isn't a versioned rockspec published to LuaRocks (just a  dev one). Unless there's something I'm missing, I think there needs to be a versioned rock published for this to work through `luarocks test` automatically (e.g., 0.1.0).

I think this is a really nice solution, but I still think there could be a way for Cyan to do a bunch of this automatically too.

**Reply by Aire-One** [reply]
Thank you for the feedback @mtdowling !

I should have mentioned it earlier: busted-tl is an initial work. There are features that are not working as expected. The most annoying one is that luassert extensions are not supported by Teal right now (https://github.com/teal-language/tl/issues/532#issuecomment-2543397647). I also had issue with coverage metrics that couldn't understand the typing stuff and think records are uncovered code. 

Also, I was the only user. So I couldn't tell if it is useful to proceed with a release version, or contributing upstream.
